filters/auth/auth.go

```diff
@@ -0,0 +1,141 @@
+package auth
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	logfilter "github.com/zalando/skipper/filters/log"
+)
+
+type roleCheckType int
+
+const (
+	checkOAuthTokeninfoAnyScopes roleCheckType = iota
+	checkOAuthTokeninfoAllScopes
+	checkOAuthTokeninfoAnyKV
+	checkOAuthTokeninfoAllKV
+	checkOAuthTokenintrospectionAnyClaims
+	checkOAuthTokenintrospectionAllClaims
+	checkOAuthTokenintrospectionAnyKV
+	checkOAuthTokenintrospectionAllKV
+	checkUnknown
+)
+
+type rejectReason string
+
+const (
+	missingBearerToken rejectReason = "missing-bearer-token"
+	missingToken       rejectReason = "missing-token"
+	authServiceAccess  rejectReason = "auth-service-access"
+	invalidSub         rejectReason = "invalid-sub-in-token"
+	inactiveToken      rejectReason = "inactive-token"
+	invalidToken       rejectReason = "invalid-token"
+	invalidScope       rejectReason = "invalid-scope"
+	invalidClaim       rejectReason = "invalid-claim"
+	invalidFilter      rejectReason = "invalid-filter"
+)
+
+const (
+	AuthUnknown = "authUnknown"
+
+	authHeaderName   = "Authorization"
+	authHeaderPrefix = "Bearer "
+	// accessTokenKey defined at https://tools.ietf.org/html/rfc7662#section-2.1
+	accessTokenKey = "access_token"
+	scopeKey       = "scope"
+	uidKey         = "uid"
+)
+
+type kv map[string]string
+
+var (
+	errUnsupportedClaimSpecified     = errors.New("unsupported claim specified in filter")
+	errInvalidAuthorizationHeader    = errors.New("invalid authorization header")
+	errInvalidToken                  = errors.New("invalid token")
+	errInvalidTokenintrospectionData = errors.New("invalid tokenintrospection data")
+)
+
+func (kv kv) String() string {
+	var res []string
+	for k, v := range kv {
+		res = append(res, k, v)
+	}
+	return strings.Join(res, ",")
+}
+
+func getToken(r *http.Request) (string, error) {
+	if tok := r.FormValue(accessTokenKey); tok != "" {
```

Comment:
i don't think we can use FormValue in a filter, because it consumes the request body, which in most cases should be forwarded to the backend.

Reply:
@aryszka It creates a separate request, so why should this matter?

Reply:
i see it here that it uses the incoming request from the filter context: https://github.com/zalando/skipper/pull/638/files#diff-97a96747f71322fa840a549316da3a1bR278

Reply:
the reason why this is a problem is in the cases when someone makes a post request with a body that should be sent to the backend service, and puts the auth token in the authorization header. What would happen in this case is that the FormValue consumes the incoming request body, and then the proxy won't be able to send it to the backend.